Sources/_StringProcessing/Utility/RegexFactory.swift

```diff
@@ -0,0 +1,193 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2021-2022 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+@available(SwiftStdlib 5.7, *)
+public enum _RegexFactory {
+  @available(SwiftStdlib 5.7, *)
+  public static func accumulate<Output>(
+    _ left: some RegexComponent,
+    _ right: some RegexComponent
+  ) -> Regex<Output> {
+    .init(node: left.regex.root.appending(right.regex.root))
+  }
+  
+  @available(SwiftStdlib 5.7, *)
+  public static func accumulateAlternation<Output>(
+    _ left: some RegexComponent,
+    _ right: some RegexComponent
+  ) -> Regex<Output> {
+    .init(node: left.regex.root.appendingAlternationCase(right.regex.root))
+  }
+  
+  @_spi(RegexBuilder)
+  @available(SwiftStdlib 5.7, *)
+  public static func assertion<Output>(
+    _ kind: DSLTree._AST.AssertionKind
+  ) -> Regex<Output> {
+    .init(node: .atom(.assertion(kind)))
+  }
+  
+  @_spi(RegexBuilder)
+  @available(SwiftStdlib 5.7, *)
+  public static func empty<Output>() -> Regex<Output> {
+    .init(node: .empty)
+  }
+  
+  @_spi(RegexBuilder)
+  @available(SwiftStdlib 5.7, *)
+  public static func scalar<Output>(
+    _ scalar: Unicode.Scalar
+  ) -> Regex<Output> {
+    .init(node: .atom(.scalar(scalar)))
+  }
+  
+  @_spi(RegexBuilder)
+  @available(SwiftStdlib 5.7, *)
+  public static func char<Output>(
+    _ char: Character
+  ) -> Regex<Output> {
+    .init(node: .atom(.char(char)))
+  }
+  
+  @_spi(RegexBuilder)
+  @available(SwiftStdlib 5.7, *)
+  public static func symbolicReference<Output>(
+    _ reference: ReferenceID
+  ) -> Regex<Output> {
+    .init(node: .atom(.symbolicReference(reference)))
+  }
+  
+  @_spi(RegexBuilder)
+  @available(SwiftStdlib 5.7, *)
+  public static func customCharacterClass<Output>(
+    _ ccc: DSLTree.CustomCharacterClass
+  ) -> Regex<Output> {
+    .init(node: .customCharacterClass(ccc))
+  }
+  
+  @available(SwiftStdlib 5.7, *)
+  public static func zeroOrOne<Output>(
+    _ component: some RegexComponent,
+    _ behavior: RegexRepetitionBehavior? = nil
+  ) -> Regex<Output> {
+    let kind: DSLTree.QuantificationKind = behavior.map { .explicit($0.dslTreeKind) } ?? .default
```

Comment:
Minor code cleanup: what about an initializer that takes an optional behavior? Then this can just be `.init(behavior)` for these call sites.